sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -81,13 +84,157 @@ private DoFnSignatures() {}
     return signature;
   }
 
+  /**
+   * The global context for the class, for use in analysis.
+   *
+   * <p>It contains much of the information that eventually becomes part of the {@link
+   * DoFnSignature}, but in an intermediate state.
+   */
+  @AutoValue
+  abstract static class GlobalAnalysisContext {
```

Comment:
Fair point. I tended towards immutable because the general case of "stuff declared in scopes" is much less error prone with immutable structures that can be extended easily without destroying the original object.

Also, my first draft had only one kind of context, so it was vital to be able to extend the context within a method without altering the global context. But we are sort of a special case here, largely because I factored the global context from the method context.

I'll try switching to mutable and see how it comes out. We can always go back to the full version if things get more complex.